src/formats/pnm.c

```diff
@@ -0,0 +1,287 @@
+                                   : f->width * bpc * (type == PNM_PGM ? 1 : 3);
+    if (it->end < it->pos + f->height * rowsz)
+        return PNM_EEOF;
+    for (size_t y = 0; y < f->height; ++y) {
+        argb_t* dst = f->data + y * f->width;
+        const uint8_t* src = it->pos + y * rowsz;
+        for (size_t x = 0; x < f->width; ++x) {
+            argb_t pix = ARGB_SET_A(0xff);
+            if (type == PNM_PBM) {
+                int bit = (src[x / 8] >> (7 - x % 8)) & 1;
+                pix |= bit - 1;
+            } else if (type == PNM_PGM) {
+                int v;
+                if (bpc == 1)
+                    v = src[x];
+                else
+                    v = src[x] << 8 | src[x + 1];
+                if (v > maxval)
+                    return PNM_EOVF;
+                if (maxval != UINT8_MAX)
+                    v = div_near(v * UINT8_MAX, maxval);
+                pix |= ARGB_SET_R(v) | ARGB_SET_G(v) | ARGB_SET_B(v);
+            } else {
+                int r, g, b;
+                if (bpc == 1) {
+                    r = src[x * 3];
+                    g = src[x * 3 + 1];
+                    b = src[x * 3 + 2];
+                } else {
+                    r = src[x * 3] << 8 | src[x * 3 + 1];
+                    g = src[x * 3 + 2] << 8 | src[x * 3 + 3];
+                    b = src[x * 3 + 4] << 8 | src[x * 3 + 5];
+                }
+                if (r > maxval || g > maxval || b > maxval)
+                    return PNM_EOVF;
+                if (maxval != UINT8_MAX) {
+                    r = div_near(r * UINT8_MAX, maxval);
+                    g = div_near(g * UINT8_MAX, maxval);
+                    b = div_near(b * UINT8_MAX, maxval);
+                }
+                pix |= ARGB_SET_R(r) | ARGB_SET_G(g) | ARGB_SET_B(b);
+            }
+            dst[x] = pix;
+        }
+    }
+    return 0;
+}
+
+enum loader_status decode_pnm(struct image* ctx, const uint8_t* data,
+                              size_t size)
+{
+    struct pnm_iter it[1];
+    bool plain;
+    int type, width, height, maxval, ret;
+
+    if (size < 2 || data[0] != 'P') {
+        return ldr_unsupported;
+    }
+    switch (data[1]) {
+        case '1':
+            plain = true;
+            type = PNM_PBM;
+            break;
+        case '2':
+            plain = true;
+            type = PNM_PGM;
+            break;
+        case '3':
+            plain = true;
+            type = PNM_PPM;
+            break;
+        case '4':
+            plain = false;
+            type = PNM_PBM;
+            break;
+        case '5':
+            plain = false;
+            type = PNM_PGM;
+            break;
+        case '6':
+            plain = false;
+            type = PNM_PPM;
+            break;
+        default:
+            return ldr_unsupported;
+    }
+    it->pos = data + 2;
+    it->end = data + size;
+
+    width = pnm_readint(it);
+    if (width < 0) {
+        image_print_error(ctx, pnm_strerror(width));
```

Comment:
```
../src/formats/pnm.c: In function ‘decode_pnm’:
../src/formats/pnm.c:243:9: warning: format not a string literal and no format arguments [-Wformat-security]
  243 |         image_print_error(ctx, pnm_strerror(width));
      |         ^~~~~~~~~~~~~~~~~
```

Use instead:
```
image_print_error(ctx, "%s", pnm_strerror(width));
```

Reply:
Sorry, I wasn't getting this warning, so didn't know I should look at it. I've now compiled with `gcc` and `clang` for both build type `debug` and `release` with no warnings, so hopefully we're in the clear.